File: python/stack.py
#!/usr/bin/env python3
# coding=utf-8

#########################################################################
#   Copyright (C) 2019 All rights reserved.
#   
#   FILE: stack.py
#   AUTHOR: Max Xu
#   MAIL: xuhuan@live.cn
#   DATE: 2019.04.22 16:45:34
#
#########################################################################


import os
import sys

class Node():
    def __init__(self, data, next=None):
        self._data = data
        self._next = next


class LinkedStack():
    def __init__(self):
        self._top: Node = None

    def push(self, data):
        print("Push", data)
        new_top = Node(data)
        new_top._next = self._top
        self._top = new_top
    
    def pop(self) -> object:
        if self._top:
            data = self._top._data
            self._top = self._top._next
            print("Pop:", data)
            return data
        else:
            return None

    def __repr__(self) -> str:
        s = ""
        top = self._top
        while top != None:
            s += "->" + str(top._data)
            print(s)
            top = top._next
        return s 


def main(args):
    ls = LinkedStack()
    ls.push(12)
    ls.push(14)
    ls.push(16)
    print(ls)

    print(ls.pop())
    print(ls.pop())
    print(ls.pop())
    print(ls.pop())


if __name__ == "__main__":
    main(sys.argv[1:])

File: go/stack.go
package main

import (
	"fmt"
	"log"
)


type LinearStack struct {
	Data []int
}

func NewLinearStack() *LinearStack {
	ls := &LinearStack{make([]int, 0, 32)}
	return ls
}

func (this *LinearStack) Push(data int) {
	log.Println("Push:", data)
	this.Data = append(this.Data, data)
}

func (this *LinearStack) Pop() interface{} {
	if len(this.Data) != 0{
		d := this.Data[len(this.Data) - 1]
		this.Data = this.Data[:len(this.Data) - 1]
		log.Println("Pop:", d)
		return d
	}
	return nil
}

func (this *LinearStack) Print() {
	if len(this.Data) == 0 {
		fmt.Println("Empty Stack.")
	} else {
		for i := len(this.Data) - 1; i >= 0; i-- {
			fmt.Print("->", this.Data[i])
		}
		fmt.Println()
	}
}


type Node struct {
	Data interface{}
	Next *Node
}

type LinkedStack struct {
	Top *Node
}

func NewLinkedStack() *LinkedStack {
	ls := &LinkedStack{nil}
	return ls
}

func (this *LinkedStack) Push(data interface{}) {
	log.Println("Push", data)
	newTop := &Node{data, nil}
	newTop.Next = this.Top
	this.Top = newTop
}

func (this *LinkedStack) Pop() interface{} {
	if this.Top != nil {
		data := this.Top.Data
		this.Top = this.Top.Next
	    log.Println("Pop", data)
		return data
	}
	return nil
}

func (this *LinkedStack) Print() {
	if this.Top == nil {
		fmt.Println("Empty Stack.")
	} else {
		top := this.Top
		s := ""
		for top != nil {
			s += "->" + fmt.Sprintf("%d", top.Data)
			top = top.Next
		}
		fmt.Println(s)
	}
}


func main() {
	fmt.Println("LinearStack")
	s := NewLinearStack()
	s.Push(12)
	s.Push(14)
	s.Push(16)
	s.Print()
	s.Pop()
	s.Pop()
	s.Pop()
	s.Pop()
	s.Print()


	fmt.Println("\nLinkedStack")
	ls := NewLinkedStack()
	ls.Push(12)
	ls.Push(14)
	ls.Push(16)
	ls.Print()
	ls.Pop()
	ls.Pop()
	ls.Pop()
	ls.Print()
}
